etcd3/client.py

```diff
@@ -126,6 +174,131 @@ def _get_secure_creds(self, ca_cert, cert_key, cert_cert):
                         cert_cert_file.read()
                     )
 
+    def _discover(self, srv_name):
+        answers = dns.resolver.query(srv_name, 'SRV')
+        hosts = {}
+        for answer in answers:
+            node = Node(
+                answer.target.to_text(omit_final_dot=True),
+                answer.port,
+                self.uses_secure_channel
+            )
+            hosts[node.netloc] = node
+        if not len(hosts):
+            raise exceptions.NoServerAvailableError(
+                "No host were found in the SRV query for {name}".format(
+                    name=srv_name
+                )
+            )
+        return hosts
+
+    @property
+    def current_node(self):
+        if self._current_node is None:
+            return None
+        return self._nodes_cache[self._current_node]
+
+    @property
+    def channel(self):
+        """
+        Get an available channel on the first node that is not failed.
+
+        Raises an exception if no node is available.
+        """
+        # No reconnection is allowed, just use the current node.
+        if not self.reconnect:
+            return self.current_node.channel
+
+        # If we already picked a node and it didn't fail,
+        # return its channel
+        if self._current_node and not self.current_node.failed:
+            return self.current_node.use()
+        else:
+            available = [m for m in self._nodes_cache.values()
+                         if not m.failed]
+            for node in available:
+                # A node might have failed in the meanwhile
+                if node.failed:
+                    next
+                self._current_node = node.netloc
+                return self.current_node.use()
+        # No node is available, so raise an exception
+        raise exceptions.NoServerAvailableError("No nodes available.")
+
+    @property
+    def kvstub(self):
+        return etcdrpc.KVStub(self.channel)
+
+    @property
+    def clusterstub(self):
+        return etcdrpc.ClusterStub(self.channel)
+
+    @property
+    def leasestub(self):
+        return etcdrpc.LeaseStub(self.channel)
+
+    @property
+    def maintenancestub(self):
+        return etcdrpc.MaintenanceStub(self.channel)
+
+    @property
+    def watchstub(self):
+        return etcdrpc.WatchStub(self.channel)
+
+    def _get_cluster_topology(self):
```

Comment:
1 - I don't really see why it's a problem, the whole point of that method is to mess with _nodes_cache, but I'm ok with doing it within the caller.

2 - In python-etcd (the v2 client) we ran a similar function every time we had one connection failing:  we would fail the node, try the call again to another node, and if successful, also upgrade the cluster topology to check if it changed. I thought it was too much to add that as well in a single patch, and there are some drawbacks from that approach too.